### PR TITLE
Use the scatter pattern to send different microbatches to different trainer DP groups

### PR DESCRIPTION
**Pull request by pzhan9**
Currently controller sends all the microbatches to all trainer ranks, and let the rank keeps what it needs, and throw the rest away. This is not efficient. This diff introduces a scatter pattern so controller will send different microbatches to different DP groups.

I also plan to use how we reshape the mesh into ("batch", "within_batch") on generator, after generator is enabled with DP. With the reshaping, it would be easy to route among different DP groups within the same generator.

### Test plan

Run `rl_grpo_qwen3_1_7b` with DP=2 to confirm it is still working:

```
python -m torchtitan.experiments.rl.train --module rl --config rl_grpo_qwen3_1_7b --metrics.no-enable-wandb --trainer.parallelism.data_parallel_shard_degree=2 --trainer.parallelism.tensor_parallel_degree=2
```

### CODE REVIEW

File: torchtitan/experiments/rl/trainer.py [hunk 1]
```diff
@@ -367,6 +360,27 @@ async def setup_async(
...
+                    for dp_rank in range(self.trainer_dp_degree)
+                ]
+            )
+            for dp_rank, value_mesh in enumerate(verifications):
+                actual = list(value_mesh.values())
+                if any(reported != dp_rank for reported in actual):
+                    raise ValueError(
+                        "Trainer mesh reshape is inconsistent with the DeviceMesh "
+                        f"batch axis: expected all to be {dp_rank}, but got {actual}."
+                    )
```
Comment (felipemello1) [inline torchtitan/experiments/rl/trainer.py]: is this something you want to land or just some temporary debugging?

**Review by felipemello1 — COMMENTED** [review]
Thanks for the PR. I am curious how this compares vs sending all batches to rank 0 and relying on torch to scatter. Do you have an intuition for it? Should we measure perf?

The positive side in my opinion is that this is a step in the direction of merging our trainer actor with titan's core actor, since the forward_backward expects the batch for that dp rank: https://github.com/pytorch/torchtitan/blob/main/torchtitan/trainer.py#L796-L800

File: torchtitan/experiments/rl/trainer.py (hunk 1 above)
Comment (pzhan9) [reply torchtitan/experiments/rl/trainer.py]: I think it is worth to keep it for the correctness sake. Otherwise we do not have anything to guarantee that we are sending to the dp ranks as we expect. It is a one-time check at the startup time, so it is pretty cheap to run.

**Comment by pzhan9** [comment]
> I am curious how this compares vs sending all batches to rank 0 and relying on torch to scatter. Do you have an intuition for it? Should we measure perf?

Intuitively, I feel this should be faster:

* In this approach, the latency is broadcasting `1/DP` microbatches.
* In the rank 0 approach, the latency is broadcasting all microbatches (which is already larger than above), plus redistribution. 

It is getting harder to say if we start to further optimize the data path. But, I think this approach definitely gives us the most ROI at the moment. (it is not a lot of code if we exclude the `verify_dp_rank` part).

The main reason drives me to submit this PR is the potential parity with generator. For generator, we would also need to do the reshaping so router can be aware of DP groups in order to route among them.

File: torchtitan/experiments/rl/trainer.py [hunk 2]
```diff
@@ -723,11 +737,16 @@ async def train(self):
...
+                    # Scatter microbatches by the DP dimension.
+                    shard_results = await asyncio.gather(
+                        *[
+                            self.trainer.slice(batch=dp_rank).forward_backward.call(
+                                microbatch[dp_rank], num_global_valid_tokens
+                            )
+                            for dp_rank in range(self.trainer_dp_degree)
+                        ]
                     )
+                    mb_metrics = self._get_rank_0_value(shard_results[0])
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/trainer.py]: could you explain & add comments to this
- what is the range of "rank_0"
- what is the range of shard_results[0] indexing

File: torchtitan/experiments/rl/trainer.py [hunk 3]
```diff
@@ -723,11 +737,16 @@ async def train(self):
             fwd_bwd_metrics: dict[str, float] = {}
             for microbatch in microbatches:
                 with sl.log_trace_span("trainer_forward_backward_call"):
-                    mb_metrics = self._get_rank_0_value(
-                        await self.trainer.forward_backward.call(
-                            microbatch, num_global_valid_tokens
-                        )
+                    # Scatter microbatches by the DP dimension.
+                    shard_results = await asyncio.gather(
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/trainer.py]: do you mean
```suggestion
                    per_dp_rank_results = await asyncio.gather(
```

File: torchtitan/experiments/rl/trainer.py [hunk 4]
```diff
@@ -367,6 +360,27 @@ async def setup_async(
                 generators.append(generator)
             self.generator_router = config.generator_router.build(generators=generators)
 
+        # Reshape trainer mesh into ("batch", "within_batch") so controller can
+        # scatter microbatches by the DP dimension.
+        self.trainer = self.trainer.flatten("flat").split(
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/trainer.py]: is "flat" a required or random name? Can we not give it a name, e.g. `flatten()`

File: torchtitan/experiments/rl/trainer.py [hunk 5]
```diff
@@ -367,6 +360,27 @@ async def setup_async(
...
+            flat=("batch", "within_batch"), batch=self.trainer_dp_degree
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/trainer.py]: if `within_batch` never used
```suggestion
            flat=("batch", "_"), batch=self.trainer_dp_degree
```

File: torchtitan/experiments/rl/actors/trainer.py [hunk 6]
```diff
@@ -340,6 +340,24 @@ async def sync_log_step(self, step: int, relative_step: int | None = None) -> No
         """Sync the structured-logger step counter from the controller."""
         sl.set_step(step, relative_step=relative_step)
 
+    @endpoint
+    async def verify_dp_rank(self, expected_dp_rank: int) -> int:
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/actors/trainer.py]: let's have more clear function & variable names, and improve docstring to reflect 
- self.dp_rank is from spmd torch distributed device mesh
- expected_dp_rank is from monarch controller mesh

File: torchtitan/experiments/rl/trainer.py (hunk 5 above)
Comment (tianyu-l) [inline torchtitan/experiments/rl/trainer.py]: "batch" is good name, as it matches the spmd device mesh naming https://github.com/pytorch/torchtitan/blob/main/torchtitan/distributed/parallel_dims.py#L281

However, the code will be wrong if PP > 1 is used in trainer, as PP is being used as the outermost mesh axis.

I'm not sure if PP has been E2E tested in RL (e.g. torchstore composability), but at least add assertion (e.g. in post_init that trainer PP == 1), and TODO.

cc @wwwjn

File: torchtitan/experiments/rl/trainer.py (hunk 4 above)
Comment (pzhan9) [reply torchtitan/experiments/rl/trainer.py]: A random name [is required](https://github.com/meta-pytorch/monarch/blob/46347a1bf84646c9293c0b4ef62ce532a771e31d/python/monarch/_src/actor/shape.py#L173) by this Monarch API.

I renamed it to "rank", which might be a better name than "flat"?

File: torchtitan/experiments/rl/trainer.py (hunk 5 above)
Comment (pzhan9) [reply torchtitan/experiments/rl/trainer.py]: Added an assert in `post_init`.

File: torchtitan/experiments/rl/trainer.py [hunk 7]
```diff
@@ -153,6 +153,20 @@ def __post_init__(self):
                     "Set generator.checkpoint.enable=False."
                 )
 
+            # The controller scatters microbatches by reshaping the trainer mesh
+            # into ("batch", "_") and slicing the "batch" axis. This assumes
+            # "batch" is the outermost mesh axis, which only holds when pipeline
+            # parallelism is off.
+            # TODO: support trainer PP by accounting for the outermost "pp" axis
+            # in the reshape and validating torchstore composability with PP.
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/trainer.py:161]: Conceptually there should be no blocker. Practically I'm not sure torchstore and monarch can already handle this. @wwwjn please investigate CP and PP trainer parallelism and raise gaps if there are any.

File: torchtitan/experiments/rl/trainer.py [hunk 8]
```diff
@@ -747,11 +777,25 @@ async def train(self):
             fwd_bwd_metrics: dict[str, float] = {}
             for microbatch in microbatches:
                 with sl.log_trace_span("trainer_forward_backward_call"):
-                    mb_metrics = self._get_rank_0_value(
-                        await self.trainer.forward_backward.call(
-                            microbatch, num_global_valid_tokens
-                        )
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/trainer.py:753]: (Prior to this PR) Letting monarch send every bit of data to all DP/TP ranks (and then perform sharding to keep necessary subset) is clearly not optimal.

There seem to be several ways to improve
1. controller scatters the microbatches to all trainer DP ranks (this PR) and there's no intra-spmd redistribute.
2. controller sends to spmd rank 0, and let it broadcast / scatter to other TP / DP (respectively) ranks in the SPMD world, via NCCL
3. controller sends to TP=0 ranks or DP=0 ranks, and let each TP / DP master do the broadcasting / scattering.

Among these options, is there a framework to analyze trade-off? This PR is doing option 1, but we've seen some other framework doing 2.

File: torchtitan/experiments/rl/trainer.py (hunk 7 above)
Comment (wwwjn) [reply torchtitan/experiments/rl/trainer.py:161]: I tried running CP PP in trainer and they can run out of pocket without changing in Monarch an torchstore. We need to add some code in policy trainer to support PP and CP.

Launched sucessful runs on: 
- Trainer: DP=2, TP=2, PP=2, CP=2
- Trainer allocation: 16 GPUs across two hosts
- Generators: two independent TP=4 meshes
